analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarReportingContextBase.cs

```diff
@@ -0,0 +1,55 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer;
+
+public abstract class SonarReportingContextBase<TContext> : SonarAnalysisContextBase<TContext>
+{
+    private protected abstract ReportingContext CreateReportingContext(Diagnostic diagnostic);
+
+    protected SonarReportingContextBase(SonarAnalysisContext analysisContext, TContext context) : base(analysisContext, context) { }
+
+    public void ReportIssue(Diagnostic diagnostic)  // FIXME: Make this obsolete
+    {
+        if (diagnostic.Descriptor.HasMatchingScope(Compilation, IsTestProject(), IsScannerRun()))
```

Comment:
Don't we check this already at the registration time?

Reply:
Yes, and we need to do it twice.  This PR preserves that logic from the past.

One analyzer class can contain two rule IDs, one has scope "Main", other has scope "All".
Rule needs to be registered => will execute => will report for both issues.
And we need to stop reporting of the "Main" one on the test code.